### DISCUSSION TITLE
`media-playlist` element

### DISCUSSION THREAD

**Discussion by mmcc** [muxinc/media-chrome#97] [Proposals]
**Note**: This isn't a Super Formal proposal, this is intended to be a "hey what if" and a test of using discussions for it 😅 

The other day I needed to create a playlist, and it made me yearn for something like the [`videojs-playlist`](https://github.com/brightcove/videojs-playlist) functionality. There would need to be quite a bit of thought around what the UI would look like for showing the playlist items, but to keep things simple assume this proposal is starting at just an auto-advancing playlist of media items.

For the most simple use-case, my thought is that we can start with 4 new elements:
## `<media-playlist>`
The parent element of all the playlist items (`<media-playlist-item>`). This element would also be where the playlist behavior was configured, such as autoadvancing, looping, pre-fetching behavior, etc.

## `<media-playlist-item>`
This gets a little weird, admittedly. In my mind this is _mostly_ the same as a `<source>` tag, but should also accept things like `poster`.

## `<media-playlist-previous-button>`
Button to be used in the `control-bar` for returning to the previous playlist item.

## `<media-playlist-next-button>`
Button to be used in the `control-bar` for advancing to the next playlist item.

## Full example
```html
<media-controller>
  <video
    slot="media"
    crossorigin
  >
    <track label="thumbnails" default kind="metadata" src="https://image.mux.com/DS00Spx1CV902MCtPj5WknGlR102V5HFkDe/storyboard.vtt">
  </video>
  <media-playlist autoadvance>
    <media-playlist-item poster="https://image.mux.com/DS00Spx1CV902MCtPj5WknGlR102V5HFkDe/thumbnail.jpg" src="https://stream.mux.com/DS00Spx1CV902MCtPj5WknGlR102V5HFkDe"></media-playlist-item>
  <media-playlist>
  <media-control-bar>
    <media-play-button></media-play-button>
    <media-playlist-previous-button><media-playlist-previous-button>
    <media-playlist-next-button><media-playlist-next-button>
  </media-control-bar>
</media-controller>
```

**Comment by cjpillsbury** [comment]
High level I think this is 🔥. Just trying to think through some things. Since `<media-playlist/>` and `<media-playlist-item/>` (or wherever the dust settles on this) will be the first non-visual component added to Media Chrome, I'm wondering:

1. How should it be represented on the DOM (e.g. is it a "peer" of our visual elements, is it the sort of thing that should be "slotted", etc)
2. How much of the data relationships/state management should be owned by media-controller "under the hood" (vs. e.g. deferring more responsibility to the JS managed by media-playlist... worrying about a ballooning set of "non-core functionality" growing in media-controller)
3. How tied do we want our implementation of the data relationships/state management of a "playlist" to media-chrome?

At a high level, we could think of this as 2 different pieces: 1) A data contract; 2) A UI that "knows about" and is responsive to/can provide updates via (1). `<media-playlist/>` & `<media-playlist-item/>` in the example above are a coarse-grained proposal for (1), whereas `<media-playlist-previous-button/>` & `<media-playlist-next-button/>` are about (2) and shouldn't require much/any discussion, aside from how/what they update "under the hood". In fact, as long as you're squinting, I don't see (1) as that different in kind from our discussion of coming up with a spec+data contract for exposing the different playlists/variants/renditions/representations (depending on your jargon) for e.g. bitrate selection, alternate camera renditions, and the like, where media-chrome "respects" and potentially "detects" the presence of the contact and also provides UI components that receive & update state via that contract, but media-chrome wouldn't be the full/sole owner of the implementation of said contract.

I have some additional thoughts/could throw out a few other straw men proposals, but wanted to start with some seed planting.

**Comment by heff** [comment]
I have a slightly different proposal, for the media-playlist-items specifically. Instead just use media elements as the items, and media-playlist becomes a custom media element itself, encapsulating the logic of rotating medias internally.

```html
<media-controller>
  <media-playlist slot="media">
    <video src="...">
    <audio src="..." preload="none">
    <youtube-video src="..." preload="none">
  </media-playlist>
</media-controller>
```

This approach gets around any need for the playlist to understand differences between custom media elements. e.g. how do I handle a youtube source or text tracks?

[x-playist](https://github.com/heff/x-playlist) is a similar demo I built years ago and could be a starting place.

The alt to that approach is to just build `<media-playlist>` into media-controller/container, so that any time you have multiple elements using slot="media", it rotates through them. Media-container is already set up to handle multiple media slots, it just doesn't do anything with that today. My worry with this version though is bloating media-container or creating unexpected behavior. I think I prefer the encapsulation of a `<media-playlist>`.

```html
<media-controller>
    <video src="..." slot="media">
    <audio src="..." slot="media" preload="none">
    <youtube-video src="..." slot="media" preload="none">
</media-controller>
```

With either, we'll also need to standardize some events, like `MEDIA_NEXT_REQUEST` and `MEDIA_PREVIOUS_REQUEST`.

We might also consider the fact that the youtube player can go through a playlist itself, so a "next()" API might need to become a [media-ui-extension](https://github.com/video-dev/media-ui-extensions).

**Reply by cjpillsbury** [reply]
> ```html
> <media-controller>
>   <media-playlist slot="media">
>     <video src="...">
>     <audio src="..." preload="none">
>     <youtube-video src="..." preload="none">
>   </media-playlist>
> </media-controller>
> ```

I like this generally, though it does assume that every custom element supports `preload="none"`. Similarly, there may still be a decent amount of work being done by the various components simply by adding them to the DOM. Since we'd want to support different elements regardless (even if in the end it doesn't make sense to slot them to `<media-playlist/>` via the default slot), I say we run with it and can always revisit if we find there are pains along the way.

The primary reason I prefer this over making them direct descendants is it's more explicit/less presumptuous about what "several slotted media elements means" (playlist? multi-player? other?). There is the potential of a "3rd way" which would likely put less effort/complexity in implementing `<media-playlist/>`. Something like:

```html
<media-controller>
  <media-playlist>
    <video slot="media" src="...">
    <audio src="..." preload="none">
    <youtube-video src="..." preload="none">
  </media-playlist>
</media-controller>
```

Where `<media-playlist/>` is responsible for updating which element is "slotted". For this @heff we'd need to confirm/revisit some of the logic in `media-container` for slotted media updates iirc.

**Reply by mmcc** [reply]
Yep, the lack of control over what exactly each media element is going to do by just existing in the DOM at load is why I leaned away from that approach initially. It could be a safe assumption that every blessed media slot is going to behave as we'd expect around things like `preload`, but that feels pretty questionable to me.

That being said, I definitely didn't account for non-`video` items in my proposal. If we were to go with a custom playlist-item approach I think we'd at least need an attribute to specify what element should be used when adding it to the DOM.

**Reply by heff** [reply]
> There is the potential of a "3rd way" which would likely put less effort/complexity in implementing <media-playlist/>.

That's an interesting approach. It feels simple on the surface but it does add complexity to media-controller understanding what media-playlist is. e.g. a `next` request would go to media-playlist while `play` goes to the media element. I'll say I was thinking it'd be cool if `<media-playlsit>` could stand on its own for other use cases, but that's not a hard requirement.

> I think we'd at least need an attribute to specify what element should be used when adding it to the DOM.

I could get behind that, to get around preload issues. It also kind of opens the door for easy metadata extensions. Something like `<mux-video>`'s `playback-id` attribute could be an issue.

```html
<media-playlist-item
  type="mux-video" <!-- matches media element name -->
  src="?"
  title="My great video"
>
```

**Reply by cjpillsbury** [reply]
> That's an interesting approach. It feels simple on the surface but it does add complexity to media-controller understanding what media-playlist is. e.g. a next request would go to media-playlist while play goes to the media element.

Unless I'm missing something, `<media-playlist/>` could work very similar to other Media State Receivers instead of thinking of it as a "special kind of slotted media":
1. The `<media-controller/>` instance receives e.g. a `MEDIA_NEXT_REQUEST`
2. It then provides an update to any Media State Receivers that are observing for the corresponding state update (`NEXT_REQUESTED`?)
3. `<media-playlist/>` would have the corresponding state listed in its `observedAttributes`, just like other components
4. It swaps out the currently slotted `media` element (either using `<media-playlist-item/>` to create the next one or using its list of children, if we got that path).
5. If desired, the `<media-playlist/>` can clear its state or emit another event representing that it's transitioned to the next/previous request.

> ```type="mux-video" <!-- matches media element name -->```

This is a good web-components analog to similar & common conventions in other languages/frameworks that support IoC for passing in a "renderer" or similar with lazy/conditional rendering 👌. 

> It also kind of opens the door for easy metadata extensions. Something like <mux-video>'s playback-id attribute could be an issue.
I'd recommend holding off on the metadata side for now, especially if the attributes specific to the media-playlist-item, the attributes that are intended as pass throughs to the media element, and the attributes that represent metadata are all flattened (as there might be name conflicts/overlaps, which I suspect you're calling out w/e.g. the playback-id concern). If we want to expand data sets, we always have the option of adding generic and/or well-defined data nodes/custom elements that are not visually rendered, e.g.:

```html
<media-playlist-item
  type="mux-video" <!-- matches media element name -->
  src="?"
  title="My great video"
>
  <metadata-attrs title="My great video"/>
</media-playlist-item>
```

**Reply by heff** [reply]
> NEXT_REQUESTED

Sure, that follows the same communication pattern, but not the same architecture. Creating `-requested` attributes and `SUCCESSFULLY_UPDATED` (or whatever) events turns a State Receiver into a State Owner, and decentralizes some of the globally relevant state. I don't think that's what we want to do, but we should at least be cautious about creating that pattern.

I want to dig into the `<youtube-video>` case a little more, because if it can support playlist-like functionality on its own, then we'd be duplicating effort supporting that and a media-playlist component that's not a media element itself.

**Reply by cjpillsbury** [reply]
Isn't that exactly what `MEDIA_REQUEST_PREVIEW` does?

```js
      MEDIA_PREVIEW_REQUEST: (e) => {
        const media = this.media;
        // No media (yet), so bail early
        if (!media) return;

        
        const [track] = getTextTracksList(media, { kind: TextTrackKinds.METADATA, label: 'thumbnails' });
        // No thumbnails track (yet) or no cues available in thumbnails track, so bail early.
        if (!(track && track.cues)) return;
        
        const time = e.detail;
        const cue = Array.prototype.find.call(track.cues, c => c.startTime >= time);

        // No corresponding cue, so bail early
        if (!cue) return;

        // Since this isn't really "global state", we may want to consider moving this "down" to the component level,
        // probably pulled out into its own module/set of functions (CJP)
        const url = new URL(cue.text);
        const previewCoordsStr = new URLSearchParams(url.hash).get('#xywh');
        this.propagateMediaState(MediaUIAttributes.MEDIA_PREVIEW_IMAGE, url.href);
        this.propagateMediaState(MediaUIAttributes.MEDIA_PREVIEW_COORDS, previewCoordsStr.split(',').join(' '));
      },
```

(from https://github.com/muxinc/media-chrome/blob/main/src/js/media-controller.js#L142-L164)

**Reply by heff** [reply]
Not exactly, but fair point. Nothing else in the system relies on the state of a preview component. We can build multiple components using the preview-related events, and it won't cause any conflicts between those components or anywhere in the system. They're also dependent on centralized state, the preview thumbnail track.

I can't say the same about a media-playlist. If there were two playlists, and something needs to know the "current-playlist-item", which is it? If one playlist successfully advances and the other doesn't, what happens? Unless I'm missing a detail of the approach, dependency on the state and results of media-playlist operations then require a way for media-playlist to inform media-controller of state in the same way media elements inform media-controller. If we use events for that, they're ultimately system-driven events, not user-driven like the existing `MediaUIEvents`.

Is complexity the only reason we wouldn't go the path of making media-playlist a media-element? I don't think it'd be nothing, but we can follow the patterns of custom-video or even subclass to make it not crazy.

**Reply by cjpillsbury** [reply]
I'm most concerned about (1) complexity and (2) introducing an additional point of failure for data flow. Each time we implement a component that "wraps" another component, we have another place where we need to maintain parity, which becomes more complex as we extend beyond the `HTMLMediaElement` definition (See, e.g., our conversation about adding support for exposing quality levels/"playlists"/"media representations" as just one example)

**Reply by cjpillsbury** [reply]
If we're ok with accepting that cost of maintenance/risk (which will require a "swiss army knife" approach to anything that fits within this architectural pattern), then that's fine, but just want to highlight it as a risk & cost.
